Fix DefaultTabController created with no Tabs

matthew5025 opened the pull request:
Prevents an assert error when the DefaultTabController is created and contains no tabs.

## Description

This PR fixes a problem when a DefaultTabController is created with no tabs.

## Related Issues

https://github.com/flutter/flutter/issues/37139

## Tests

I added the following tests:

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

matthew5025 committed:
Fix DefaultTabController

Prevents an assert error when the DefaultTabController is created and contains no tabs.

HansMuller commented:
This appears to be a safe change however it will need a test.

CC @shihaohong

matthew5025 commented:
Apologies @HansMuller , but this is my first PR and I'll be happy to include a test but I'm not very sure what sort of test to include.

shihaohong commented on packages/flutter/lib/src/material/tab_controller.dart:
```diff
@@ -313,7 +313,7 @@ class DefaultTabController extends StatefulWidget {
   }) : assert(initialIndex != null),
        assert(length >= 0),
-       assert(initialIndex >= 0 && initialIndex < length),
+       assert(initialIndex >= 0 && initialIndex <= length),
```
I'm unsure if this is actually a safe change, since for all other cases, `initialIndex` can now be set to a value outside of the range of tab values (e.g. Having two tabs and having `initialIndex` set to 2 will no longer trigger this assert), This will probably be better:
```suggestion
       assert(length == 0 || (initialIndex >= 0 && initialIndex <= length)),
```

HansMuller replied:
Yes! Good point.

matthew5025 replied:
@shihaohong Thanks for the feedback! If I've understood your point correctly, the code should be changed to this instead?

```suggestion
              assert(length == 0 || (initialIndex >= 0 && initialIndex < length)),
```

shihaohong replied:
Yes

shihaohong reviewed (changes requested):
I have a comment about updating this assert statement to something that makes a little more sense. As for the test, I would recommend maybe adding a test case where you have zero tabs and proving that the assertion passes.

Go into `${FLUTTER_HOME}/packages/flutter/test/material/tabs_test.dart` to see other examples of tests and then add one that will fail without the fix and pass after your fix has been included. The sample that causes triggers the error would be a great starting point for how to design that test case

matthew5025 committed:
Update packages/flutter/lib/src/material/tab_controller.dart

Fixed index / length assert check

matthew5025 committed:
Added test for a DefaultTabController with no tabs

The test creates a DefaultTabController with zero tabs, add one tab and deletes it again.

matthew5025 commented:
> I have a comment about updating this assert statement to something that makes a little more sense. As for the test, I would recommend maybe adding a test case where you have zero tabs and proving that the assertion passes.
> 
> Go into `${FLUTTER_HOME}/packages/flutter/test/material/tabs_test.dart` to see other examples of tests and then add one that will fail without the fix and pass after your fix has been included. The sample that causes triggers the error would be a great starting point for how to design that test case

@shihaohong I've added a test; would appreciate it it you could have a look at it. Thanks!

matthew5025 committed:
Fixed prefer const issue with CI test

matthew5025 committed:
Removed whitespace at the end of line

matthew5025 committed:
Fixed more whitespace issues

shihaohong commented on packages/flutter/test/material/tabs_test.dart:
```diff
@@ -2350,4 +2350,94 @@ void main() {
+    testWidgets('DefaultTabController with zero tabs',
+          (WidgetTester tester) async {
+        // Regression test for https://github.com/flutter/flutter/issues/32428
+        int noOfTabs = 0;
```
```suggestion
```

shihaohong commented on packages/flutter/test/material/tabs_test.dart:
```diff
@@ -2350,4 +2350,94 @@ void main() {
+            home: StatefulBuilder(
+              builder: (BuildContext context, StateSetter setState) {
+                return DefaultTabController(
+                    length: noOfTabs,
```
```suggestion
                    length: tabList.length,
```

shihaohong commented on packages/flutter/test/material/tabs_test.dart:
```diff
@@ -2350,4 +2350,94 @@ void main() {
+                    child: Scaffold(
+                      appBar: AppBar(
+                        title: const Text('Default TabBar Preview'),
+                        bottom: (noOfTabs > 0)
```
```suggestion
                        bottom: tabList.isNotEmpty
```

shihaohong commented on packages/flutter/test/material/tabs_test.dart:
```diff
@@ -2350,4 +2350,94 @@ void main() {
+                            ? TabBar(isScrollable: true, tabs: tabList)
+                            : null,
+                      ),
+                      body: (noOfTabs > 0)
```
```suggestion
                      body: tabList.isNotEmpty
```

shihaohong commented on packages/flutter/test/material/tabs_test.dart:
```diff
@@ -2350,4 +2350,94 @@ void main() {
+                            IconButton(
+                              key: const Key('Add tab'),
+                              icon: const Icon(Icons.add),
+                              onPressed: () {
```
This logic can be further simpified if you're simply adding a new item to your list:
```
                          onPressed: () {
                            setState(() {
                              tabList.add(Tab(text: 'Tab ${tabList.length + 1}'));
                              tabViewList.add(Center(child: Text('Tab ${tabViewList.length + 1}')));
                            });
                          },
```

shihaohong replied:
Also, it's better practice to have your state change logic happen within `setState`

shihaohong commented on packages/flutter/test/material/tabs_test.dart:
```diff
@@ -2350,4 +2350,94 @@ void main() {
+                            IconButton(
+                              key: const Key('Del tab'),
+                              icon: const Icon(Icons.delete),
+                              onPressed: () {
```
ditto:
```
                          onPressed: () {
                            setState(() {
                              tabList.removeLast();
                              tabViewList.removeLast();
                            });
                          },
```

shihaohong commented on packages/flutter/test/material/tabs_test.dart:
```diff
@@ -2350,4 +2350,94 @@ void main() {
     expect(find.text('Tab2'), findsOneWidget);
   });
+
+    testWidgets('DefaultTabController with zero tabs',
```
It also seems that the indentations here are off by 2 spaces compared to the rest of the file.

shihaohong replied:
It's better to specify here what you mean, since this can be misunderstood as the `TabBar.tabs` or the `TabBarView.children` list
```suggestion
    testWidgets('DefaultTabController should allow for a length of zero',
```

shihaohong commented on packages/flutter/test/material/tabs_test.dart:
```diff
@@ -2350,4 +2350,94 @@ void main() {
+                          ],
+                        ),
+                      ),
+                    ));
```
Fix the closing brackets here

shihaohong reviewed (changes requested):
I have some feedback regarding the formatting of your code and simplifying the logic of updating tab state. Otherwise, the change looks good! Let me know if you have any questions.

matthew5025 committed:
Updates tabs_test as per suggestions

matthew5025 commented:
@shihaohong Thanks for the suggestions, I've updated the test file as per your feedback.

On an (somewhat) unrelated note, I've noticed that the new test code works when you add one tab, but adding the second tab fails.
Is this an expected behaviour or did I muck something up?
The error is
`The following RangeError was thrown building TabBar(dirty, dependencies: [_TabControllerScope, _LocalizationsScope-[GlobalKey#ba4d0], Directionality, _InheritedTheme], state: _TabBarState#f149b):
RangeError (index): Invalid value: Only valid value is 0: 1
`

The code for the simple app I've used is [here](https://gist.github.com/matthew5025/0b29d62bf274f7ed75c7379171e9a406)

matthew5025 committed:
Updated tab_test to use isNotEmpty

matthew5025 commented:
@shihaohong Apologies for the disturbance; is there any updates on this?

shihaohong commented:
Hi, sorry for the long delay! I'll review this later today and investigate the case where you add a second tab.

shihaohong commented on packages/flutter/test/material/tabs_test.dart:
```diff
@@ -2350,4 +2350,72 @@ void main() {
     expect(find.text('Tab1'), findsOneWidget);
     expect(find.text('Tab2'), findsOneWidget);
   });
+  testWidgets('DefaultTabController should allow for a length of zero', (
```
```suggestion
  testWidgets('DefaultTabController should allow for a length of zero', (WidgetTester tester) async {
```

shihaohong commented on packages/flutter/test/material/tabs_test.dart:
```diff
@@ -2350,4 +2350,72 @@ void main() {
     expect(find.text('Tab2'), findsOneWidget);
   });
+  testWidgets('DefaultTabController should allow for a length of zero', (
+      WidgetTester tester) async {
```
```suggestion
```

shihaohong commented on packages/flutter/test/material/tabs_test.dart:
```diff
@@ -2350,4 +2350,72 @@ void main() {
   });
+  testWidgets('DefaultTabController should allow for a length of zero', (
+      WidgetTester tester) async {
+    // Test for https://github.com/flutter/flutter/issues/37139
```
```suggestion
```
Since this is not a regression, this comment is not needed.

shihaohong commented on packages/flutter/test/material/tabs_test.dart:
```diff
@@ -2350,4 +2350,72 @@ void main() {
+  testWidgets('DefaultTabController should allow for a length of zero', (
+      WidgetTester tester) async {
+    // Test for https://github.com/flutter/flutter/issues/37139
+    final List<Widget> tabList = <Widget>[];
```
I suggested this at the beginning, but realized we could do better here. Instead of using two separate lists of widgets, we could use a List<String> to define both the TabBar and TabView contents instead:
```suggestion
  List<String> tabTextContent = <String>[];
```

shihaohong commented on packages/flutter/test/material/tabs_test.dart:
```diff
@@ -2350,4 +2350,72 @@ void main() {
+    final List<Widget> tabViewList = <Widget>[];
+
+    await tester.pumpWidget(
+      MaterialApp(
```
Here's my suggestion, let me know what you think:
```
    MaterialApp(
      home: StatefulBuilder(
        builder: (BuildContext context, StateSetter setState) {
          return DefaultTabController(
            length: tabTextContent.length,
            child: Scaffold(
              appBar: AppBar(
                title: const Text('Default TabBar Preview'),
                bottom: tabTextContent.isNotEmpty
                  ? TabBar(
                    isScrollable: true,
                    tabs: tabTextContent.map((textContent) {
                      return Tab(text: textContent);
                    }).toList(),
                  )
                  : null,
              ),
              body: tabTextContent.isNotEmpty
                ? TabBarView(
                  children: tabTextContent.map((textContent) {
                    return Tab(text: '$textContent\'s view');
                  }).toList())
                : const Center(child: Text('No tabs')),
              bottomNavigationBar: BottomAppBar(
                child: Row(
                  mainAxisSize: MainAxisSize.max,
                  mainAxisAlignment: MainAxisAlignment.spaceBetween,
                  children: <Widget>[
                    IconButton(
                      icon: const Icon(Icons.add),
                      onPressed: () {
                        setState(() {
                          tabTextContent = new List.from(tabTextContent)
                            ..add('Tab ${tabTextContent.length + 1}');
                        });
                      },
                    ),
                    IconButton(
                      icon: const Icon(Icons.delete),
                      onPressed: () {
                        setState(() {
                          tabTextContent = new List.from(tabTextContent)
                            ..removeLast();
                        });
                      },
                    ),
                  ],
                ),
              ),
            ),
          );
        },
      ),
    );
```

shihaohong commented on packages/flutter/test/material/tabs_test.dart:
```diff
@@ -2350,4 +2350,72 @@ void main() {
+    expect(find.text('No tabs'), findsOneWidget);
+    await tester.tap(find.byKey(const Key('Add tab')));
+    await tester.pumpAndSettle();
+    expect(find.text('Tab 1'), findsNWidgets(2));
```
I modified the `TabBarView`'s text in my example to allow for checking both in isolation, since that is probably much more readable to future maintainers. So, check for "Tab {number}'s view" in `TabBarView`, and also check for "Tab {number}" in `TabBar`

shihaohong reviewed (changes requested):
Here are my suggestions for improving the test, as well as properly addressing the change in tab length that's causing the error case you mentioned. Thanks for being thorough!

shihaohong commented:
This has been fixed with https://github.com/flutter/flutter/pull/41625. Thank you so much for your initial contributions!

matthew5025 commented:
@shihaohong Apologies for the lack of response; been very busy with school these days. Also, thanks for following up on this and fixing it!
